Modules/_sqlite/connection.c

```diff
@@ -897,21 +897,122 @@ print_or_clear_traceback(callback_context *ctx)
+    if (str == NULL) {
+        return NULL;
+    }
+
+    /* Store a strong reference in the module state, and a borrowed
+     * reference to the caller. */
+    state->bltin_str = Py_NewRef(str);
+    return str;
+}
+
+static PyObject *
+get_exception_message(callback_context *ctx, PyObject *exc)
+{
+    PyObject *str = fetch_str_builtin(ctx);  // Borrowed ref.
+    if (str == NULL) {
+        return NULL;
+    }
+    PyObject *res = PyObject_CallFunction(str, "(O)", exc);
+    if (res == NULL) {
+        return NULL;
+    }
+    return res;
+}
+
+static char *
+build_error_message(const char *preamble, const char *message)
+{
+    assert(preamble || message);
+    size_t n = 0;
+    n += preamble ? strlen(preamble) : 0;
+    n += message ? strlen(message) : 0;
+    if (preamble && message) {
+        n += 2;  // Make room for ": ".
+    }
+    char *buf = PyMem_Malloc(n + 1);
+    if (buf == NULL) {
+        return (char *)PyErr_NoMemory();
+    }
+    buf[n] = 0;
+    if (preamble && !message) {
+        return strncpy(buf, preamble, n);
+    }
+    if (!preamble && message) {
+        return strncpy(buf, message, n);
+    }
+    buf = strncpy(buf, preamble, n);
+    buf = strncat(buf, ": ", n);
+    return strncat(buf, message, n);
+}
+
+static char *
+get_result_error_message(callback_context *ctx, const char *preamble,
+                         PyObject *exc)
+{
+    PyObject *msg_obj = get_exception_message(ctx, exc);
+    if (msg_obj == NULL) {
+        return NULL;
+    }
+    const char *msg = PyUnicode_AsUTF8(msg_obj);
+    if (msg == NULL) {
+        Py_DECREF(msg_obj);
+        return NULL;
+    }
+    char *errmsg = build_error_message(preamble, msg);
+    Py_DECREF(msg_obj);
+    return errmsg;
+}
+
 // Checks the Python exception and sets the appropriate SQLite error code.
 static void
-set_sqlite_error(sqlite3_context *context, const char *msg)
+set_sqlite_error(sqlite3_context *context, const char *preamble)
 {
     assert(PyErr_Occurred());
+    callback_context *ctx = (callback_context *)sqlite3_user_data(context);
     if (PyErr_ExceptionMatches(PyExc_MemoryError)) {
         sqlite3_result_error_nomem(context);
     }
     else if (PyErr_ExceptionMatches(PyExc_OverflowError)) {
         sqlite3_result_error_toobig(context);
     }
     else {
-        sqlite3_result_error(context, msg, -1);
+        PyObject *exc = PyErr_GetRaisedException();
+
+        char *msg = get_result_error_message(ctx, preamble, exc);
+        if (msg == NULL) {
+            // An exception happened while we tried to build a more detailed
+            // error message; write this as an unraisable exception and pass
+            // the preamble as the sqlite3 error message.
+            void *data = sqlite3_user_data(context);
+            callback_context *ctx = (callback_context *)data;
+            PyErr_WriteUnraisable(ctx->callable);
+            sqlite3_result_error(context, preamble, -1);
+        }
```

Comment:
We should be able to test this, for example like this:

<details>

```diff
diff --git a/Lib/test/test_sqlite3/test_userfunctions.py b/Lib/test/test_sqlite3/test_userfunctions.py
index 67a13ad38b..d8b8c0a40e 100644
--- a/Lib/test/test_sqlite3/test_userfunctions.py
+++ b/Lib/test/test_sqlite3/test_userfunctions.py
@@ -24,6 +24,7 @@
 import sys
 import unittest
 import sqlite3 as sqlite
+import test.support
 
 from unittest.mock import Mock, patch
 from test.support import bigmemtest, gc_collect
@@ -35,6 +36,10 @@
 class CustomException(Exception):
     pass
 
+class EvilException(Exception):
+    def __str__(self):
+        return -1
+
 def func_returntext():
     return "foo"
 def func_returntextwithnull():
@@ -53,6 +58,8 @@ def func_returnlonglong():
     return 1<<31
 def func_raiseexception():
     raise CustomException("func except")
+def func_raiseevil():
+    raise EvilException("my __str__() does not return a string")
 def func_memoryerror():
     raise MemoryError
 def func_overflowerror():
@@ -162,6 +169,7 @@ def setUp(self):
         self.con.create_function("return_noncont_blob", 0,
                                  lambda: memoryview(b"blob")[::2])
         self.con.create_function("raiseexception", 0, func_raiseexception)
+        self.con.create_function("raiseevil", 0, func_raiseevil)
         self.con.create_function("memoryerror", 0, func_memoryerror)
         self.con.create_function("overflowerror", 0, func_overflowerror)
 
@@ -265,6 +273,18 @@ def test_func_exception(self):
             cur.execute("select raiseexception()")
             cur.fetchone()
 
+    @with_tracebacks(EvilException, name="func_raiseevil")
+    def test_func_evilexception(self):
+        cur = self.con.cursor()
+        # Since it is impossible to fetch a string with the exception message
+        # from EvilException, we will only receive the generic preamble
+        # exception message, hence the detailed regex.
+        msg = "^user-defined function raised exception$"
+        with self.assertRaisesRegex(sqlite.OperationalError, msg):
+            with test.support.catch_unraisable_exception() as cm:
+                cur.execute("select raiseevil()")
+                cur.fetchone()
+
     @with_tracebacks(MemoryError, name="func_memoryerror")
```

</details>

Unfortunately, our `@with_tracebacks` decorator gets in the way. I'll think of a way to solve that (we can also handle it later in a follow-up PR).